### PR TITLE
Category and offer scraping

### PR DESCRIPTION
**Pull request by PLJohnny**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -98,4 +98,4 @@ ENV/
 /site
 
 # mypy
-.mypy_cache/
+.mypy_cache/
```
Comment (PythonicNinja) [inline .gitignore:103]: new line at the end of file

File: otodom.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    except AttributeError as e:
+        total_floors = ""
+    return total_floors
+
+
+def get_offer_apartment_details(html_parser):
+    try:
+        found = html_parser.find(class_="sub-list")
+    except AttributeError as e:
+        return {}
```
Comment (PythonicNinja) [inline otodom.py]: maybe found = False

File: otodom.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    return details
+
+
+def get_offer_additional_assets(html_parser):
+    try:
+        found = html_parser.findAll(class_="dotted-list")
+    except AttributeError as e:
+        return []
+    asset_list = sum([str(detail.text).split("\n") for detail in found], [])
+    assets = []
```
Comment (PythonicNinja) [inline otodom.py]: list of comprehansion?

File: otodom.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+
+def get_offer_description(html_parser):
+    description = html_parser.find(itemprop="description").text.replace(u'\xa0', u' ').replace('\n', '')
+    return description
+
+
+def get_offer_photos_links(html_parser):
+    gallery = html_parser.findAll(class_="gallery-box-thumb-item")
+    photo_links = []
```
Comment (PythonicNinja) [inline otodom.py]: list of comprehansions

File: otodom.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    try:
+        found = html_parser.find(class_="sub-list")
+    except AttributeError as e:
+        return {}
+    details = {}
+    if found:
+        detail_list = str(found.text).split("\n")
+    else:
+        return {}
+    for detail in detail_list:
```
Comment (PythonicNinja) [inline otodom.py]: dict of comprehension?

**Comment by PLJohnny** [comment]
I wrongly assumed that scraping all the apartments in a given city would be enough, but the truth is, there's also stuff like districts and streets that should be taken into consideration when filtering. I have to rework the get_city_from_autosuggest method to return all the necessary values.

File: otodom.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    offers = html_parser.find_all(class_="offer-item")
+    parsed_offers = [parse_category_offer(str(offer)) for offer in offers if offer]
+    return parsed_offers
+
+
+def get_category_number_of_pages(markup):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    offers = html_parser.find(class_="current")
+    try:
+        return int(offers.contents[0])
```
Comment (PythonicNinja) [inline otodom.py]: maybe `int(offers.text)`

File: otodom.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+    except KeyError as e:
+        # offer was not present any more
+        return {}
+
+    number_replace_dict = {u'\xa0': "", " ": "", "-": "", "+48": ""}
+    phone_numbers = sum([replace_all(num, number_replace_dict).split(".") for num in phone_numbers], [])
+
+    facebook_description = get_offer_facebook_description(html_parser)
+
+    ninja_pv = get_offer_ninja_pv(content)
```
Comment (PythonicNinja) [inline otodom.py]: maybe we can make it a bit more compact?

File: otodom.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+
+def get_offer_photos_links(html_parser):
+    gallery = html_parser.findAll(class_="gallery-box-thumb-item")
+    return [p.attrs["href"] for p in gallery]
+
+
+def get_offer_geographical_coordinates(html_parser):
+    latitude = html_parser.find(itemprop="latitude").attrs["content"]
+    longitude = html_parser.find(itemprop="longitude").attrs["content"]
+    return [latitude, longitude]
```
Comment (PythonicNinja) [inline otodom.py]: `return latitude, longitude` it will be a tuple

File: otodom.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        return {}
+    return details
+
+
+def get_offer_additional_assets(html_parser):
+    try:
+        found = html_parser.findAll(class_="dotted-list")
+        assets = [a for a in sum([str(detail.text).split("\n") for detail in found], []) if a]
+    except AttributeError as e:
+        return []
```
Comment (PythonicNinja) [inline otodom.py]: or `assets = []`

File: otodom.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+        total_floors = ""
+    return total_floors
+
+
+def get_offer_apartment_details(html_parser):
+    try:
+        found = html_parser.find(class_="sub-list")
+        details = [{d.split(": ")[0]: d.split(": ")[1]} for d in str(found.text).split("\n") if d]
+    except AttributeError as e:
+        return {}
```
Comment (PythonicNinja) [inline otodom.py]: `details = {}`
one return is easier to wrap head around

File: otodom.py [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+    city_or_voivodeship = region_data["city"] if "city" in region_data else region_data["region"]
+
+    if "[district_id]" in region_data:
+        filters["[district_id]"] = region_data["[district_id]"]
+
+    # creating base url
+    url = "/".join([BASE_URL, main_category, detail_category, city_or_voivodeship])
+
+    # adding building type if exists in filters
+    try:
```
Comment (PythonicNinja) [inline otodom.py]:
```
if "building_type" in filters:
     url = url + "/" + filters["building_type"]
```

File: otodom.py [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+        url = url + "/" + filters["building_type"]
+    except KeyError as e:
+        pass
+
+    # adding description fragment search if exists in filters
+    try:
```
Comment (PythonicNinja) [inline otodom.py]: if "building_type" in filters:
     url = url + "/" + filters["building_type"]

File: otodom.py [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+    filter_list = []
+    for key, value in filters.items():
+        if isinstance(value, list):
+            for item in value:
+                filter_list.append("search{}={}".format(quote(key), item))
+        else:
+            filter_list.append("search{}={}".format(quote(key), value))
+
+    url = url + "&".join([ads_per_page, page, *filter_list])
+    print(url)
```
Comment (PythonicNinja) [inline otodom.py]: removal of that print, maybe configure logging?
usage:
`log.info(url)`

File: requirements.txt [hunk 14]
```diff
@@ -0,0 +1,4 @@
+requests
+beautifulsoup4
+https://github.com/limebrains/scrapper-helpers/archive/master.zip
+re
```
Comment (PythonicNinja) [inline requirements.txt]: re is builtin

File: otodom.py [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+def get_category_number_of_pages(markup):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    offers = html_parser.find(class_="current")
+    try:
+        return int(offers.text)
+    except AttributeError as e:
+        return 1
+
+
+def was_search_successful(markup):
```
Comment (PythonicNinja) [inline otodom.py]: `was_category_search_successful`

File: otodom.py [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+
+
+def get_offer_geographical_coordinates(html_parser):
+    latitude = html_parser.find(itemprop="latitude").attrs["content"]
+    longitude = html_parser.find(itemprop="longitude").attrs["content"]
+    return [latitude, longitude]
+
+
+def get_offer_details(html_parser):
+    f = html_parser.find(class_="text-details clearfix").text
```
Comment (PythonicNinja) [inline otodom.py]: maybe remove `clearfix`?

File: otodom.py [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+        district = ninja_pv["district_name"]
+    except KeyError as e:
+        district = ""
+    surface = ninja_pv["surface"]
+    rooms = ninja_pv["rooms"]
+    poster_type = ninja_pv["poster_type"]
+
+    floor = get_offer_floor(html_parser)
+    total_floors = get_offer_total_floors(html_parser)
+
```
Comment (PythonicNinja) [inline otodom.py]: maybe if it can be done for all attrs in response dict:
```
{
'surface': ninja_pv["surface"],
...
}
```

File: otodom.py [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+        region_dict["city"] = unicodedata.normalize('NFKD', text[1].lower()).encode('ascii', 'ignore').decode("utf-8")
+        region_dict["[district_id]"] = response["district_id"]
+    elif type == "REGION":
+        region_dict["region"] = unicodedata.normalize('NFKD', text[0].lower()).encode('ascii', 'ignore').decode("utf-8")
+
+    return region_dict
+
+
+def get_url(main_category, detail_category, region, ads_per_page="", page=None, **filters):
+    if page != None:
```
Comment (PythonicNinja) [inline otodom.py]: `is not None`

File: otodom.py [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+        page = "page=" + str(page)
```
Comment (PythonicNinja) [inline otodom.py]: `page = "page={0}".format(str(page)) if page is not None else "page"`

File: otodom.py [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+    phone_numbers = sum([replace_all(num, number_replace_dict).split(".") for num in phone_numbers], [])
+
+    facebook_description = get_offer_facebook_description(html_parser)
+
+    ninja_pv = get_offer_ninja_pv(content)
+    price = ninja_pv["ad_price"]
+    currency = ninja_pv["price_currency"]
+    voivodeship = ninja_pv["region_name"]
+    city = ninja_pv["city_name"]
+    try:
```
Comment (PythonicNinja) [inline otodom.py]: `if "district_name" in ninja_pv`

Comment (PythonicNinja) [reply otodom.py]: or even better `ninja_pv.get("district_name", "")`

File: otodom.py [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+    for i, j in dic.items():
+        text = text.replace(i, j)
+    return text
+
+
+def get_region_from_autosuggest(region_part):
+    url = "https://www.otodom.pl/ajax/geo6/autosuggest/?data={0}".format(region_part)
+    response = json.loads(get_response_for_url(url).text)[0]
+    type = response["level"]
+    text = response["text"].replace("<strong>", "").replace("</strong>", "").split(", ")
```
Comment (PythonicNinja) [inline otodom.py]: `city, district = response["text"].replace("<strong>", "").replace("</strong>", "").split(", ")`

than in later occurances:
replace `text[0].lower()` with `city.lower()`

File: otodom.py [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+def replace_all(text, dic):
+    for i, j in dic.items():
+        text = text.replace(i, j)
+    return text
+
+
+def get_region_from_autosuggest(region_part):
+    url = "https://www.otodom.pl/ajax/geo6/autosuggest/?data={0}".format(region_part)
+    response = json.loads(get_response_for_url(url).text)[0]
+    type = response["level"]
```
Comment (PythonicNinja) [inline otodom.py]: type is builtin
if you really need that name, there is convention of adding `type_` to not override builtin variables.
But maybe changing name would be better.

File: otodom.py [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+    offer_id = context['offer_id']
+
+    # getting offer details
+    try:
+        phone_numbers = get_offer_phone_numbers(offer_id, cookie, csrf_token)
+    except KeyError as e:
+        # offer was not present any more
+        return {}
+
+    number_replace_dict = {u'\xa0': "", " ": "", "-": "", "+48": ""}
```
Comment (PythonicNinja) [inline otodom.py]: `phone_number_replace_dict`

File: otodom.py [hunk 24]
```diff
@@ -0,0 +1,362 @@
...
+            **context,
+        }
+    }
+
+
+if __name__ == '__main__':
+
+    # the following commented dict contains every possible filter with examples of its values
+
+    # input_dict = {
```
Comment (PythonicNinja) [inline otodom.py]: maybe move it to get_category docstring[1]


[1] - https://www.python.org/dev/peps/pep-0257/

File: otodom.py (hunk 21 above)
Comment (PLJohnny) [reply otodom.py]: The position of city in this list depends on the "level". If it's "CITY", city is in text[0], if it's "DISTRICT", city is in text[1]. Using variables here would only add to the confusion

File: otodom.py [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+    logging.basicConfig(level=logging.INFO)
+
+BASE_URL = 'http://www.otodom.pl'
+SCRAPE_LIMIT = os.environ.get('SCRAPE_LIMIT', None)
+
+WHITELISTED_DOMAINS = [
+    'otodom.pl',
+    'www.otodom.pl',
+]
+
```
Comment (PythonicNinja) [inline otodom.py:17]: we could if DEBUG log config

File: scrape/category.py [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+        'detail_url': url,
+        'offer_id': offer_id,
+        'poster': poster.text.strip() if poster else "",
+    }
+
+
+def parse_category_content(markup):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    offers = html_parser.find_all(class_="offer-item")
+    parsed_offers = [parse_category_offer(str(offer)) for offer in offers if
```
Comment (PythonicNinja) [inline scrape/category.py]: new line just after first `[`
```
parsed_offers = [
    parse_category_offer(str(offer)) for offer in offers
    if offer.attrs["data-featured-name"] != "promo_vip" and
     offer.attrs["data-featured-name"] != "promo_top_ads"
]
```

File: scrape/category.py [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+    return parsed_offers
+
+
+def get_category_number_of_pages(markup):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    offers = html_parser.find(class_="current")
+    try:
+        return int(offers.text)
+    except AttributeError as e:
+        return 1
```
Comment (PythonicNinja) [inline scrape/category.py]: maybe `log.exception`

File: scrape/offer.py [hunk 28]
```diff
@@ -0,0 +1,177 @@
...
+    return title
+
+
+def get_offer_address(html_parser):
+    address = html_parser.find(class_="address-text").text
+    return address
+
+
+def get_offer_information(url, context=None):
+    # geting response
```
Comment (PythonicNinja) [inline scrape/offer.py]: `getting`

File: scrape/utils.py [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+        region_dict["city"] = "{0}{1}{2}".format(unicodedata.normalize('NFKD', text[0].lower()).encode('ascii', 'ignore').decode("utf-8").replace(" ", "-"), "_", response["city_id"])
+        region_dict["[street_id]"] = response["street_id"]
+
+    return region_dict
+
+
+def get_url(main_category, detail_category, region, ads_per_page="", page=None, **filters):
+    page = "page={0}".format(page) if page is not None else ""
+
+    region_data = get_region_from_autosuggest(region)
```
Comment (PythonicNinja) [inline scrape/utils.py]: maybe check if user already added it in filters skip autosugest?

File: tests.py [hunk 30]
```diff
@@ -0,0 +1,127 @@
+import pytest
+import pickle
+from bs4 import BeautifulSoup
+from unittest import mock
```
Comment (PythonicNinja) [inline tests.py:4]: it's python module so it should be at the very top

File: tests.py [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+#             get_region_from_autosuggest.return_value = get_region_from_autosuggest_value
+#             url = utils.get_url(main_category, detail_category, region)
+
+@pytest.mark.parametrize(
+    'markup_file,expected_value', [
+        ("test_data/markup_no_offers", "95c3ad18f9c716209fdfc5d73b13f4a64fd12fc7ca9b6d0d4a5f60ce80b574b3"),
+        ("test_data/markup_offers", "d6e9f6202c0fd68ddc539a54bd728d59aa27d7276470818a57ed7c3c2db5f612")
+    ])
+def test_get_csrf_token(markup_file, expected_value):
+    with open(markup_file, "rb") as file:
```
Comment (PythonicNinja) [inline tests.py]: file is builtin name use sth else,
maybe `markup_file` could be `markup_path` and then as `markup_file`?

File: scrape/category.py [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def was_category_search_successful(markup):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    return False if html_parser.find(class_="search-location-extended-warning") else True
```
Comment (PythonicNinja) [inline scrape/category.py]:
```
has_warning = bool(html_parser.find(class_="search-location-extended-warning"))
return not has_warning
```

File: scrape/utils.py [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+def get_region_from_autosuggest(region_part):
+    if not region_part:
+        return {}
+    url = "https://www.otodom.pl/ajax/geo6/autosuggest/?data={0}".format(region_part)
+    response = json.loads(get_response_for_url(url).text)[0]
+    region_type = response["level"]
+    text = response["text"].replace("<strong>", "").replace("</strong>", "").split(", ")
+
+    region_dict = {}
+
```
Comment (PythonicNinja) [inline scrape/utils.py:60]:
```
lambda input_text: "{0}{1}{2}".format(unicodedata.normalize('NFKD', input_text.lower()).encode('ascii', 'ignore').decode("utf-8").replace(" ", "-"), "_", response["city_id"])
```

File: .gitignore (hunk 1 above)
Comment (PythonicNinja) [reply .gitignore:103]: @PLJohnny

File: scrape/category.py [hunk 34]
```diff
@@ -0,0 +1,122 @@
...
+        offer.attrs["data-featured-name"] != "promo_top_ads"
+    ]
+    return parsed_offers
+
+
+def get_category_number_of_pages(markup):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    offers = html_parser.find(class_="current")
+    if offers:
+        return int(offers.text)
```
Comment (PythonicNinja) [inline scrape/category.py]: `return int(offers.text) if offers else 1`

File: scrape/category.py [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+        '[street_id]': from otodom API
+    }
+
+    :return: A list of dictionaries. Each of them contains the following fields:
+        'detail_url' - a link to the offer
+        'offer_id' - the internal otodom's offer ID, not to be mistaken with the '[id]' field from the input_dict
+        'poster' - a piece of information about the poster. Could either be a name of the agency or "Oferta prywatna"
+    """
+
+    url = get_url(main_category, detail_category, region, "?nrAdsPerPage=72", **filters)
```
Comment (PythonicNinja) [inline scrape/category.py]: there is probably simpler way

**Comment by PLJohnny** [comment]
I forgot about filters specific to offers for sale. [filter_float_price_per_m:from], [filter_float_price_per_m:to], [filter_enum_market][], [filter_enum_building_material][]. All I have to do is document it, though. The backend will deal with them correctly.

**Pull request merged**